### ISSUE TITLE
mpi_file_open fails on symlink with Lustre filesystem

### ISSUE THREAD

**Issue by anton-seaice** [open-mpi/ompi#12141]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

v4.1.4

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

From source tarball

### Please describe the system on which you are running

* Operating system/version: Rocky Linux 8.8 (Green Obsidian)
* Computer hardware: Intel Xeon Scalable ‘Cascade Lake’
* Network type: ? (see https://nci.org.au/our-systems/hpc-systems)

-----------------------------

## Details of the problem

When I attempt to open a file through its symlink, it fails with the error :
```
get_stripe failed: 61 (No data available)
MPI_ERR_OTHER: known error not in list
```
The expected result is that the symlink to the file opens the same way the file does.

This is my minimal test example:

```
program mpi_file_open_example
  implicit none

  include 'mpif.h'
  
  integer, parameter :: nprocs = 4
  integer :: rank, ierr, file_handle, status, len 
  character(len=MPI_MAX_ERROR_STRING) :: error_string
  character(len=80) :: filename = "parallel_file.txt", link = "link_to_parallel_file.txt"

  ! Confirm the link can be opened using open
  open(newunit=file_handle, file=link, status="old", action="read", &
  iostat=ierr, iomsg=error_string)
  if (ierr /= 0) then
    print *, trim(error_string)
    call abort()
  end if
    close(file_handle)

  
  ! Initialize MPI
  call MPI_Init(ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  ! Open the file in parallel
  write(*, *) "Open the file in parallel: "
  call MPI_File_open(MPI_COMM_WORLD, filename, &
                     MPI_MODE_RDONLY, MPI_INFO_NULL, file_handle, ierr)
    
  if (ierr /= MPI_SUCCESS) then
    call MPI_Error_string(ierr, error_string, len, ierr)
    write(*, *) "Error opening ", trim(filename), " : ", trim(error_string)
  else
    write(*, *) "Success opening ", trim(filename)
  end if
  
  ! Close the file
  call MPI_File_close(file_handle, ierr)

  write(*, *) "==================="

  ! Open the symlink in parallel
  write(*, *) "Open the symlink in parallel: "
  call MPI_File_open(MPI_COMM_WORLD, link, &
                     MPI_MODE_RDONLY, MPI_INFO_NULL, file_handle, ierr)
    
  if (ierr /= MPI_SUCCESS) then
    call MPI_Error_string(ierr, error_string, len, ierr)
    write(*, *) "Error opening ", trim(link), " : ", trim(error_string)
  else
    write(*, *) "Success opening ", trim(link)
  end if

  ! Close the file
  call MPI_File_close(file_handle, ierr)

  ! Finalize MPI
  call MPI_Finalize(ierr)
  
end program mpi_file_open_example

```

to reproduce:

```
module purge
module load intel-compiler/2021.6.0 openmpi/4.1.4
mpifort mpi_file_open.f90 -o mpi_file_open.exe

touch parallel_file.txt
ln -s parallel_file.txt link_to_parallel_file.txt

./mpi_file_open.exe 
```

**Comment by edgargabriel** [comment]
@anton-seaice thank you for submitting the issue, I will try to look into this in the next few days

**Comment by anton-seaice** [comment]
Hi @edgargabriel How did you go with this?

**Comment by edgargabriel** [comment]
@anton-seaice really sorry, I didn't get to it last week. It is on my todo list for this week however.

**Comment by benmenadue** [comment]
I did some testing: `llapi_file_get_stripe` doesn't work through symlinks. Or more precisely it tries to get the striping of the symlink itself, which naturally fails as they're metadata-only entities with no objects.

**Comment by edgargabriel** [comment]
Does it work btw. if you force using the romio component, e.g. adding ```--mca io romio321``` to your mpirun command line?

**Comment by anton-seaice** [comment]
> Does it work btw. if you force using the romio component, e.g. adding `--mca io romio321` to your mpirun command line?

Thanks for the suggestion. That does work. In our model code though, the read/write timers report it being a lot slower (takes about twice as long). I will investigate some more :)

**Comment by benmenadue** [comment]
That works because romio opens the file itself and uses the `LL_IOC_LOV_GETSTRIPE` ioctl to get the striping info. The open follows the symlink and hence this works find with symlinks (and even with symlinks that aren't themselves on a Lustre filesystem but do target objects that are).

On the other hand, `llapi_file_get_stripe` opens the parent directory and uses the `IOC_MDC_GETFILESTRIPE` ioctl. Since the name of the object to query is passed as part of the ioctl, it bypasses the kernel's symlink handling and hence operates on the symlink itself instead of the target.

**Comment by edgargabriel** [comment]
@anton-seaice @benmenadue could you please test the fix in https://github.com/open-mpi/ompi/pull/12161 and let me know whether it resolves the issue? 

I tested bits and pieces of it, but I do not have a working Lustre file system at the moment for testing the full patch.
If it works, I will backport the fix to the 4.1 and the 5.0 branches

**Comment by anton-seaice** [comment]
Thanks @edgargabriel and @benmenadue 

I get these errors when I try an compile the sample code from the first post now:
```
$ mpifort mpi_file_open.f90 -o mpi_file_open.exe

/bin/ld: warning: libucc.so.1, needed by /scratch/public/openmpi/patch/lib/libmpi_usempif08.so, not found (try using -rpath or -rpath-link)
/bin/ld: warning: libpbs.so.0, needed by /scratch/public/openmpi/patch/lib/libmpi_usempif08.so, not found (try using -rpath or -rpath-link)
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_team_destroy'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_team_create_test'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_collective_finalize'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_collective_post'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_context_config_read'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_lib_get_attr'
//scratch/public/openmpi/patch/lib/libopen-rte.so.40: undefined reference to `tm_poll'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_context_config_release'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_team_create_post'
//scratch/public/openmpi/patch/lib/libopen-rte.so.40: undefined reference to `tm_init'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_lib_config_release'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_lib_config_read'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_lib_config_modify'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_init_version'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_context_destroy'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_context_create'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_finalize'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_context_progress'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_collective_init'
//scratch/public/openmpi/patch/lib/libopen-rte.so.40: undefined reference to `tm_spawn'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_context_config_modify'
/scratch/public/openmpi/patch/lib/libmpi.so: undefined reference to `ucc_status_string'
```

**Comment by benmenadue** [comment]
@anton-seaice that's my fault, I forgot to set the rpath in the libraries (because it's not through our build system I needed to set them manually)... will fix that now

**Comment by benmenadue** [comment]
@anton-seaice try it now

**Comment by anton-seaice** [comment]
Thanks @edgargabriel 

All works for me with the patched build - it opens the symlink files now.

I couldn't test the case of a symlink from a lustre to a non-lustre file system.

**Comment by anton-seaice** [comment]
Hi @edgargabriel 

Whats the timeline for the next release? Thanks

**Comment by edgargabriel** [comment]
This patch is scheduled 5.0.2, which is scheduled for around February (I think), and 4.1.7 for late Q1 2024

**Comment by wenduwan** [comment]
Patch is merged to 4.1 and 5.0 branches and will become available in the next release.
